File: Data Structures & Algorithms/longest-increasing-subsequence/submission-34.py
class Solution:
    def lengthOfLIS(self, nums: List[int]) -> int:
        
        def dfs(i, last): 
            if i == len(nums): 
                return 0 

            cur = dfs(i + 1, last) 

            if last == None or nums[i] > last: 
                cur = max(cur, 1 + dfs(i + 1, nums[i])) 

            return cur 

        return dfs(0, None)

File: Data Structures & Algorithms/longest-repeating-substring-with-replacement/submission-34.js
class Solution {
    /**
     * @param {string} s
     * @param {number} k
     * @return {number}
     */
    characterReplacement(s, k) {
        let count = {} 
        let l = 0 
        let maxLen = 0
        let maxSubstring = 0
        for (let r = 0; r < s.length; r ++) {
            let cur = s[r] 
            count[cur] = ( count[cur] || 0 ) + 1
            maxSubstring = Math.max(maxSubstring, count[cur])
            let curLen = r - l + 1 
            while(curLen - maxSubstring > k) {
                let deleted = s[l] 
                count[deleted] -- 
                l ++ 
                curLen = r - l + 1
            }

            maxLen = Math.max(maxLen, curLen)
        }

        return maxLen
    }
}
